src/serial/MidiSessionALSA.cc

```diff
@@ -166,13 +174,253 @@ void MidiOutALSA::serialize(Archive& /*ar*/, unsigned /*version*/)
+	midiConnector.setParityBit(false, SerialDataInterface::EVEN); // no parity
+
+	setConnector(&midiConnector); // base class will do this in a moment,
+	                              // but thread already needs it
+	connect();
+}
+
+void MidiInALSA::unplugHelper(EmuTime::param /*time*/)
+{
+	if (connected) {
+		disconnect();
+	}
+}
+
+void MidiInALSA::connect()
+{
+	destinationPort = snd_seq_create_simple_port(
+		&seq, "MIDI in pluggable",
+		SND_SEQ_PORT_CAP_WRITE | SND_SEQ_PORT_CAP_SUBS_WRITE, SND_SEQ_PORT_TYPE_MIDI_GENERIC);
+	if (destinationPort < 0) {
+		throw PlugException(
+			"Failed to create ALSA port: ", snd_strerror(destinationPort));
+	}
+
+	int err = snd_seq_connect_from(&seq, destinationPort, srcClient, srcPort);
+	if (err) {
+		snd_seq_delete_simple_port(&seq, destinationPort);
+		throw PlugException(
+			"Failed to connect to ALSA port "
+			"(", srcClient, ':', srcPort, ")"
+			": ", snd_strerror(err));
+	}
+
+	snd_midi_event_new(MidiOutDevice::MAX_MESSAGE_SIZE, &event_parser);
+	snd_midi_event_no_status(event_parser, 1);
+
+	connected = true;
+	stop = false;
+
+	thread = std::thread([this]() { run(); });
+}
+
+void MidiInALSA::disconnect()
+{
+	stop = true;
+	thread.join();
+
+	snd_midi_event_free(event_parser);
+	snd_seq_disconnect_from(&seq, destinationPort, srcClient, srcPort);
+	snd_seq_delete_simple_port(&seq, destinationPort);
+
+	connected = false;
+}
+
+void MidiInALSA::run()
+{
+	assert(isPluggedIn());
+
+	auto npfd = snd_seq_poll_descriptors_count(&seq, POLLIN);
+	std::vector<struct pollfd> pfd(npfd);
+	snd_seq_poll_descriptors(&seq, pfd.data(), npfd, POLLIN);
+
+	while (true) {
+		if (poll(pfd.data(), npfd, 1000) > 0) {
+			do {
+				snd_seq_event_t *ev = NULL;
+				long err, size;
+
+				if ((err = snd_seq_event_input(&seq, &ev)) < 0) {
+					std::cerr << "Error receiving MIDI event: "
+						<< snd_strerror(err) << '\n';
+					break;
+				}
+
+				if (!ev) {
+					break;
+				}
+
+				if (ev->type == SND_SEQ_EVENT_SYSEX) {
+					std::lock_guard<std::mutex> lock(mutex);
+
+					for (auto i : xrange(ev->data.ext.len)) {
+						queue.push_back(static_cast<uint8_t*>(ev->data.ext.ptr)[i]);
+					}
+				} else {
+					uint8_t bytes[12] = {};
+
+					size = snd_midi_event_decode(event_parser, bytes, 12, ev);
+					if (size < 0) {
+						std::cerr << "Error decoding MIDI event: "
+							<< snd_strerror(size) << '\n';
+						snd_seq_free_event(ev);
+						break;
+					}
+
+					std::lock_guard<std::mutex> lock(mutex);
+					for (auto i : xrange(size)) {
+						queue.push_back(bytes[i]);
+					}
+				}
+				snd_seq_free_event(ev);
+				eventDistributor.distributeEvent(
+					Event::create<MidiInALSAEvent>());
+			} while (0);
+		}
+		if (stop)
+			break;
+	}
+}
+
+void MidiInALSA::signal(EmuTime::param time)
+{
+	auto* conn = static_cast<MidiInConnector*>(getConnector());
+	if (!conn->acceptsData()) {
+		std::lock_guard<std::mutex> lock(mutex);
+		queue.clear();
+		return;
+	}
+	if (!conn->ready()) return;
+
+	byte data;
```

Comment:
Detail. I see the code uses a mix of `byte` and `uint8_t`, both are exactly the same.
A lot of code in openMSX uses a mix of these two types (we're slowly migrating to `uint8_t`). Though within one file we try to be consistent.

Reply:
Done.